### PR TITLE
cli: reindex function addition

### PR DESCRIPTION
**Pull request by hachreak**
Signed-off-by: Leonardo Rossi <leonardo.r@cern.ch>
Co-authored-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

### CODE REVIEW

**Comment by jma** [comment]
+1

**Comment by lnielsen** [comment]
Please add tests. Also I'm not unsure about naming and the fact that we adding a dependency on PIDStore.

P.S. Please add ``Co-authored-by`` as the code was taken from Zenodo ;-)

**Comment by hachreak** [comment]
I'm writing the tests.
for the `Co-authored-by`, of course! Sorry, I forgot.. :)
I'll add now.

**Comment by hachreak** [comment]
ping @jirikuncar @lnielsen do you have time to review the PR? :)

File: invenio_indexer/cli.py [hunk 1]
```diff
@@ -52,3 +53,19 @@ def run(delayed, concurrency):
...
+@with_appcontext
+def reindex(pid_type, index, doctype):
+    """Reindex all records."""
+    query = (x[0] for x in PersistentIdentifier.query.filter_by(
+        pid_type=pid_type, object_type='rec'
+    ).values(
+        PersistentIdentifier.object_uuid
+    ))
+    click.echo("Sending tasks...")
+    RecordIndexer().bulk_index(query, index=index, doc_type=doctype)
```
Comment (jma) [inline invenio_indexer/cli.py]: As I know  you should call `process_bulk_queue()` for syn call or `process_bulk_queue.delay()` for async call. Here is what I'm using actually, it is probably not optimal, but i t works for me:

        @utils.command()
        @click.argument('pid_type')
        @click.option('--bulk-size', '-s', default=50)
        @with_appcontext
        def reindex(pid_type, bulk_size):
            """Reindex using bulks."""
            query = (x[0] for x in PersistentIdentifier.query.filter_by(
                pid_type=pid_type, object_type='rec'
            ).values(
                PersistentIdentifier.object_uuid
            ))
            click.echo("Sending tasks...")

            from invenio_indexer.tasks import index_record, process_bulk_queue
            #for rec in query:
                #index_record.delay(str(rec))
                #click.echo(rec)
            to_index = []
            for rec in query:
                to_index.append(rec)
                if len(to_index) == bulk_size:
                    RecordIndexer().bulk_index(to_index)
                    process_bulk_queue.delay()
                    to_index = []
                    
            if to_index:
                click.echo("Last %d recds" % len(to_index))
                RecordIndexer().bulk_index(to_index)
                process_bulk_queue.delay()

Comment (hachreak) [inline invenio_indexer/cli.py]: Actually, I resolved the problem calling:
```bash
invenio3 index reindex recid
invenio3 index run
```

Comment (jma) [inline invenio_indexer/cli.py]: Ok but why two calls?

Comment (lnielsen) [inline invenio_indexer/cli.py]: It's correct what @hachreak is doing. The job of the CLI is only to request indexing of the given records, not to actually index them. The indexing should happen in separate celery tasks. Which you can start like this:

``<cmd> index -d --concurrency 8``

Having multiple indexing tasks running at once will get you much higher ingestion rates that what a single process can do alone.

File: .editorconfig [hunk 2]
```diff
@@ -37,7 +37,7 @@ charset = utf-8
 indent_size = 4
 # isort plugin configuration
 known_first_party = invenio_indexer
-known_third_party = invenio_db,invenio_search,invenio_records
+known_third_party = invenio_db,invenio_search,invenio_records,invenio_marc21,dojson
```
Comment (lnielsen) [inline .editorconfig]: why are these needed?

File: requirements-devel.txt [hunk 3]
```diff
@@ -24,3 +24,4 @@
 
 -e git+git://github.com/inveniosoftware/invenio-records.git#egg=invenio-records
 -e git+git://github.com/inveniosoftware/invenio-search.git#egg=invenio-search
+-e git+git://github.com/inveniosoftware/dojson.git#egg=dojson
```
Comment (lnielsen) [inline requirements-devel.txt]: ditto here, why is dojson needed?

File: invenio_indexer/cli.py [hunk 4]
```diff
@@ -52,3 +53,19 @@ def run(delayed, concurrency):
     else:
         click.secho("Indexing records...", fg='green')
         RecordIndexer().process_bulk_queue()
+
+
+@index.command()
+@click.argument('pid_type')
+@click.option('-i', '--index', default=None, help="the index name.")
```
Comment (jirikuncar) [inline invenio_indexer/cli.py]: `help='Specify the index name.'`

File: invenio_indexer/cli.py [hunk 5]
```diff
@@ -52,3 +53,19 @@ def run(delayed, concurrency):
...
+@click.option('-d', '--doctype', default=None, help="the doc_type.")
```
Comment (jirikuncar) [inline invenio_indexer/cli.py]: `--doctype` vs `--doc-type`

`help='Specify the document type.'`

File: invenio_indexer/cli.py [hunk 6]
```diff
@@ -52,3 +53,19 @@ def run(delayed, concurrency):
     else:
         click.secho("Indexing records...", fg='green')
         RecordIndexer().process_bulk_queue()
+
+
+@index.command()
```
Comment (jirikuncar) [inline invenio_indexer/cli.py]: you can try to make `reindex` and `run` chainable (http://click.pocoo.org/5/commands/#multi-command-chaining).

Comment (hachreak) [inline invenio_indexer/cli.py]: Probably to be chainable, it's needed a PR in search module, because the [@index.group()](https://github.com/inveniosoftware/invenio-search/blob/master/invenio_search/cli.py#L47) is there

File: .editorconfig [hunk 7]
```diff
@@ -37,7 +37,7 @@ charset = utf-8
 indent_size = 4
 # isort plugin configuration
 known_first_party = invenio_indexer
-known_third_party = invenio_db,invenio_search,invenio_records
+known_third_party = invenio_db,invenio_search,invenio_records,invenio_marc21
```
Comment (lnielsen) [inline .editorconfig]: Invenio-Indexer has no dependency on Invenio-MARC21 - i.e. it shouldn't be listed here.

File: tests/test_cli.py [hunk 8]
```diff
@@ -41,3 +48,33 @@ def test_run(script_info):
...
+
+        schema = {
+            'type': 'object',
+            'properties': {
+                'title': {'type': 'string'},
+                'field': {'type': 'boolean'},
+            },
+            'required': ['title'],
+        }
+        record = Record.create({'title': 'Test0', '$schema': schema})
```
Comment (lnielsen) [inline tests/test_cli.py]:
```python
rec_uuid = uuid.uuid4()
data = {'title': …, '$schema': … }
pid = recid_minter(rec_uuid, data)
Record.create(data, id_=rec_uuid)
db.session.commit()
```

File: tests/test_cli.py [hunk 9]
```diff
@@ -41,3 +48,33 @@ def test_run(script_info):
...
+        recid_minter(record.model.id, record.model.json)
+        with db.session.begin_nested():
+            db.session.add(record.model)
+        rec_id = record.id
```
Comment (lnielsen) [inline tests/test_cli.py]: Please use ``rec_uuid`` instead of ``rec_id`` as otherwise it's very easy to confuse with the Persistent Identifier recid (which is an integer) vs the record model identifier (which is a uuid)

File: invenio_indexer/api.py [hunk 10]
```diff
@@ -164,12 +164,13 @@ def delete_by_id(self, record_uuid):
         """Delete record from index by record identifier."""
         self.delete(Record.get_record(record_uuid))
 
-    def bulk_index(self, record_id_iterator):
+    def bulk_index(self, record_id_iterator, index=None, doc_type=None):
```
Comment (lnielsen) [inline invenio_indexer/api.py]: As discussed IRL, the best approach is to ensure ``$schema`` attribute is present in the record (i.e. please revert this change)

File: invenio_indexer/cli.py [hunk 11]
```diff
@@ -52,3 +53,20 @@ def run(delayed, concurrency):
...
+@click.argument('pid_type')
+@click.option('-i', '--index', default=None, help="Specify the index name.")
```
Comment (lnielsen) [inline invenio_indexer/cli.py]: Ditto here.

**Comment by jirikuncar** [comment]
Superseded by #10.
